### PR TITLE
feat: Implement pairing API

### LINKED ISSUE #13630: Implementation of `parings-api-impl` using JNI, arkworks, and alt-bn128
#### Overview

Following https://github.com/hashgraph/hedera-services/blob/8354afb91d643d8e06f5ef67d2914f032b95a5d7/platform-sdk/docs/proposals/TSS-Library/TSS-Library.md
Sections: 
- Swirlds Crypto Pairings Alt-128

Create the API interfaces and implementation to enable cryptographic pairings.

- These interfaces should provide group elements, field elements, and all required operations to support the TSS Signing scheme

#### Definitions
Repository: [hedera-cryptography](https://github.com/hashgraph/hedera-cryptography)

Gradle Modules:
- `swirlds-crypto-altbn128-impl`

Java Modules: 
- `com.swirlds.pairings.altbn128.impl`

-----

### This task depends on:
- [ ] https://github.com/hashgraph/hedera-services/issues/13324
- [ ] https://github.com/hashgraph/hedera-services/issues/14100
- [x] https://github.com/hashgraph/hedera-services/issues/14136
- [x] https://github.com/hashgraph/hedera-services/issues/14140
- [ ]  https://github.com/hashgraph/hedera-services/issues/14194
- [ ]  https://github.com/hashgraph/hedera-services/issues/14195

### PR DESCRIPTION
**Pull request by litt3**
- closes #13630

### CODE REVIEW

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/Field.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ * @see FieldElement
+ */
+public interface Field {
+    /**
+     * Creates a random field element
+     *
+     * @param random the source of randomness
+     * @return the random field element
+     */
+    default FieldElement randomElement(@NonNull final Random random) {
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/Field.java:39]:
```suggestion
    @NonNull
    default FieldElement randomElement(@NonNull final Random random) {
```

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/Curve.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        return id;
+    }
+
+    /**
+     * Get the curve from the curve id.
+     *
+     * @param curveId the curve id
+     * @return the curve
+     */
+    public static Curve fromId(final byte curveId) {
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/Curve.java:52]:
```suggestion
    @NonNull
    public static Curve fromId(final byte curveId) {
```

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/FieldElement.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    default int size() {
+        return getField().getElementSize();
+    }
+
+    /**
+     * Returns the field the element is in
+     *
+     * @return the field
+     */
+    Field getField();
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/FieldElement.java:50]:
```suggestion
    @NonNull
    Field getField();
```

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/Group.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+ */
+public interface Group {
+    /**
+     * Returns the opposite group of this group
+     * <p>
+     * If this group is G1, then the opposite group is G2, and vice versa.
+     *
+     * @return the opposite group
+     */
+    default Group getOppositeGroup() {
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/Group.java:41]:
```suggestion
    @NonNull
    default Group getOppositeGroup() {
```

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/GroupElement.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    default int size() {
+        return isCompressed() ? getGroup().getCompressedSize() : getGroup().getUncompressedSize();
+    }
+
+    /**
+     * Returns the group of the element
+     *
+     * @return the element's group
+     */
+    Group getGroup();
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/GroupElement.java:59]:
```suggestion
    @NonNull
    Group getGroup();
```

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/GroupElementAggregator.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ */
+public class GroupElementAggregator {
+    private final List<GroupElement> elements;
+
+    /**
+     * Constructor
+     */
+    public GroupElementAggregator() {
+        this.elements = new ArrayList<>();
+    }
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/GroupElementAggregator.java:36]: Does it make sense to create an empty aggregation? Do you think we can remove the option from the API?
A static builder that receives the first element as a parameter.
But also, I might be off and have to look at how you are using it.

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/GroupElementAggregator.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    public void add(@NonNull final GroupElement element) {
+        elements.add(element);
+    }
+
+    /**
+     * Compute the aggregate of the elements
+     *
+     * @return the aggregate of the elements
+     */
+    public GroupElement compute() {
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/GroupElementAggregator.java:52]:
```suggestion
    @NonNull
    public GroupElement compute() {
```

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/PairingResult.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * This class represents the result of the pairing operation between two elements
+ */
+public interface PairingResult {
+    /**
+     * Get the first input element. This element is in the opposite group of the second input element
+     *
+     * @return the first input element
+     */
+    GroupElement getInputElement1();
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/PairingResult.java:28]:
```suggestion
    @NonNull
    GroupElement getInputElement1();
```

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/PairingResult.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    /**
+     * Get the second input element. This element is in the opposite group of the first input element
+     *
+     * @return the second input element
+     */
+    GroupElement getInputElement2();
```
Comment (mxtartaglia-sl) [inline platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/PairingResult.java:35]:
```suggestion
    @NonNull
    GroupElement getInputElement2();
```

**Review by mxtartaglia-sl — COMMENTED** [review]
commenting to myself

File: platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/GroupElementAggregator.java (hunk 6 above)
Comment (alittley) [reply platform-sdk/swirlds-pairings-api/src/main/java/com/swirlds/pairings/api/GroupElementAggregator.java:36]: The problem here is that you'd need to either create an identity element as your first element (and pay the corresponding performance price).

Or alternatively mangle loops that aggregate many elements with special first-iteration logic

**Review by rbarkerSL — APPROVED** [review]
Reviewed
- platform-sdk/swirlds-pairings-api/build.gradle.kts
- .github/CODEOWNERS
- settings.gradle.kts

These files LGTM

**Review by mxtartaglia-sl — APPROVED** [review]

**Review by edward-swirldslabs — APPROVED** [review]

**Comment by mxtartaglia-sl** [comment]
Closing this pull request

**Pull request closed without merging**